deepinv/sampling/sde.py

```diff
@@ -0,0 +1,408 @@
+    :param torch.device device: device on which the computation is performed.
+    """
+
+    def __init__(
+        self,
+        drift: Callable = lambda x, t: -x,  # Default to Ornstein-Uhlenbeck process
+        diffusion: Callable = lambda t: math.sqrt(2.0),
+        use_backward_ode=False,
+        denoiser: nn.Module = None,
+        rescale: bool = False,
+        rng: torch.Generator = None,
+        dtype=torch.float64,
+        device=torch.device("cpu"),
+        *args,
+        **kwargs,
+    ):
+        super().__init__()
+        if not rescale:
+            self.denoiser = denoiser
+        else:
+            self.denoiser = (
+                lambda x, sigma, *args, **kwargs: denoiser(
+                    (x + 1) * 0.5, sigma / 2, *args, **kwargs
+                )
+                * 2.0
+                - 1.0
+            )
+        self.rng = rng
+        self.device = device
+        self.dtype = dtype
+        self.use_backward_ode = use_backward_ode
+        forward_drift = lambda x, t, *args, **kwargs: drift(x, t, *args, **kwargs)
+        forward_diff = lambda t: diffusion(t)
+        self.forward_sde = BaseSDE(
+            drift=forward_drift,
+            diffusion=forward_diff,
+            rng=rng,
+            dtype=dtype,
+            device=device,
+        )
+
+        if self.use_backward_ode:
+            backward_drift = lambda x, t, *args, **kwargs: -drift(x, t) + 0.5 * (
+                diffusion(t) ** 2
+            ) * self.score(x, t, *args, **kwargs)
+            backward_diff = lambda t: 0.0
+        else:
+            backward_drift = lambda x, t, *args, **kwargs: -drift(x, t) + (
+                diffusion(t) ** 2
+            ) * self.score(x, t, *args, **kwargs)
+            backward_diff = lambda t: diffusion(t)
+        self.backward_sde = BaseSDE(
+            drift=backward_drift,
+            diffusion=backward_diff,
+            rng=rng,
+            dtype=dtype,
+            device=device,
+        )
+
+    @torch.no_grad()
+    def forward(
+        self,
+        x_init: Union[Tensor, Tuple],
+        timesteps: Union[Tensor, List, ndarray],
+        method: str = "Euler",
+        *args,
+        **kwargs,
+    ) -> SDEOutput:
+        r"""
+        Sample the backward-SDE.
+
+        :param Union[Tensor, Tuple] x_init: Initial condition of the backward-SDE.
+            If it is a :meth:`torch.Tensor`, `x_init` should follow the distribution of :math:`p_T`, which is usually :math:`\mathcal{N}(0, \sigma_{\mathrm{max}}^2 \mathrm{Id})`.
+            If it is a tuple, it should be of the form `(B, C, H, W)`. A sample from the distribution :math:`p_T` will be generated automatically.
+
+        :param torch.Tensor timesteps: The time steps at which to discretize the backward-SDE, should be of shape `(n_steps,)`.
+        :param str method: The method to discretize the backward-SDE, can be one of the methods available in :meth:`deepinv.sampling.sde_solver`.
```

Comment:
I would move the method parameter to the class constructor and pass a Class, ie EulerSolver instead of a string

Reply:
It could be a good idea but it might be a little harder to use. We can make this similar to `https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html` or `https://github.com/google-research/torchsde`. In the future, when we add more solvers, we put other solver-related arguments to, for example, `**options`. I prefer to keep it as simple as possible, but I'am open to considering changes if necessary.

Reply:
I think it is not optimal to have multiple options that are dependent on your specific solver.

We have currently the same problem with `deepinv.optim.optim_builder` which asks for a dictionary `params_algo` whose keys can change according to the chosen algorithm - some users told me that they found this not very intuitive.

Reply:
Agree! I'll try to make the necessary changes asap.

Reply:
I just made the necessary changes